File: nodejs/docker/manual-instrumentation/express/otel-config.js
const { NodeSDK } = require('@opentelemetry/sdk-node');
const { Resource } = require('@opentelemetry/resources');
const { SemanticResourceAttributes } = require('@opentelemetry/semantic-conventions');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-http');
const { OTLPMetricExporter } = require('@opentelemetry/exporter-metrics-otlp-http');
const { OTLPLogExporter } = require('@opentelemetry/exporter-logs-otlp-http');
const { PeriodicExportingMetricReader } = require('@opentelemetry/sdk-metrics');
const { LoggerProvider, BatchLogRecordProcessor } = require('@opentelemetry/sdk-logs');
const { logs } = require('@opentelemetry/api-logs');
const { HttpInstrumentation } = require('@opentelemetry/instrumentation-http');
const { ExpressInstrumentation } = require('@opentelemetry/instrumentation-express');

function configureOpenTelemetry() {
    // Get configuration from environment or use defaults
    const serviceName = process.env.OTEL_SERVICE_NAME || 'nodejs-express-manual';
    const serviceVersion = process.env.OTEL_SERVICE_VERSION || '1.0.0';

    const tracesEndpoint = process.env.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT ||
                           process.env.OTEL_EXPORTER_OTLP_ENDPOINT ||
                           'http://localhost:4338';
    const metricsEndpoint = process.env.OTEL_EXPORTER_OTLP_METRICS_ENDPOINT ||
                            'http://localhost:4318';
    const logsEndpoint = process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT ||
                         'http://localhost:4328';

    // Create resource with service information
    const resource = Resource.default().merge(
        new Resource({
            [SemanticResourceAttributes.SERVICE_NAME]: serviceName,
            [SemanticResourceAttributes.SERVICE_VERSION]: serviceVersion,
        }),
    );

    // Configure exporters
    const traceExporter = new OTLPTraceExporter({
        url: tracesEndpoint,
    });

    const metricExporter = new OTLPMetricExporter({
        url: metricsEndpoint,
    });

    const logExporter = new OTLPLogExporter({
        url: logsEndpoint,
    });

    // Configure metrics reader
    const metricReader = new PeriodicExportingMetricReader({
        exporter: metricExporter,
        exportIntervalMillis: 60000, // Export every 60 seconds
    });

    // Configure logs
    const loggerProvider = new LoggerProvider({ resource });
    loggerProvider.addLogRecordProcessor(new BatchLogRecordProcessor(logExporter));
    logs.setGlobalLoggerProvider(loggerProvider);

    // Manual instrumentation: explicitly configure each instrumentation
    const sdk = new NodeSDK({
        resource,
        traceExporter,
        metricReader,
        loggerProvider,
        instrumentations: [
            // HTTP instrumentation for incoming/outgoing HTTP requests
            new HttpInstrumentation({
                requestHook: (span, request) => {
                    // Add custom attributes to HTTP spans
                    span.setAttribute('custom.http.instrumentation', 'manual');
                },
            }),
            // Express instrumentation for Express framework
            new ExpressInstrumentation({
                requestHook: (span, request) => {
                    // Add custom attributes to Express spans
                    span.setAttribute('custom.express.instrumentation', 'manual');
                },
            }),
        ],
    });

    // Initialize SDK
    sdk.start();

    console.log('=== OpenTelemetry Manual Instrumentation Configured ===');
    console.log(`Service Name: ${serviceName}`);
    console.log(`Service Version: ${serviceVersion}`);
    console.log(`Traces Endpoint: ${tracesEndpoint}`);
    console.log(`Metrics Endpoint: ${metricsEndpoint}`);
    console.log(`Logs Endpoint: ${logsEndpoint}`);
    console.log('=======================================================');

    // Graceful shutdown
    process.on('SIGTERM', () => {
        sdk.shutdown()
            .then(() => console.log('OpenTelemetry SDK shut down'))
            .catch((error) => console.error('Error shutting down SDK', error))
            .finally(() => process.exit(0));
    });

    return sdk;
}

module.exports = { configureOpenTelemetry };

File: python/docker/manual-instrumentation/flask/otel_config.py
import os
import logging
from opentelemetry import trace, metrics
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.exporter.otlp.proto.http._log_exporter import OTLPLogExporter
from opentelemetry.sdk._logs import LoggerProvider, LoggingHandler
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
from opentelemetry._logs import set_logger_provider
from opentelemetry.instrumentation.flask import FlaskInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor

def configure_opentelemetry():
    """Configure OpenTelemetry with manual instrumentation for Flask"""

    # Get configuration from environment or use defaults
    service_name = os.getenv('OTEL_SERVICE_NAME', 'python-flask-docker-manual')
    service_version = os.getenv('OTEL_SERVICE_VERSION', '1.0.0')

    traces_endpoint = os.getenv(
        'OTEL_EXPORTER_OTLP_TRACES_ENDPOINT',
        os.getenv('OTEL_EXPORTER_OTLP_ENDPOINT', 'http://sematext-agent:4338')
    )
    metrics_endpoint = os.getenv(
        'OTEL_EXPORTER_OTLP_METRICS_ENDPOINT',
        'http://sematext-agent:4318'
    )
    logs_endpoint = os.getenv(
        'OTEL_EXPORTER_OTLP_LOGS_ENDPOINT',
        'http://sematext-agent:4328'
    )

    # Create resource with service information
    resource = Resource.create({
        SERVICE_NAME: service_name,
        SERVICE_VERSION: service_version,
    })

    # Configure Tracing
    trace_provider = TracerProvider(resource=resource)
    trace_exporter = OTLPSpanExporter(endpoint=traces_endpoint)
    trace_provider.add_span_processor(BatchSpanProcessor(trace_exporter))
    trace.set_tracer_provider(trace_provider)

    # Configure Metrics
    metric_exporter = OTLPMetricExporter(endpoint=metrics_endpoint)
    metric_reader = PeriodicExportingMetricReader(
        exporter=metric_exporter,
        export_interval_millis=60000  # Export every 60 seconds
    )
    meter_provider = MeterProvider(
        resource=resource,
        metric_readers=[metric_reader]
    )
    metrics.set_meter_provider(meter_provider)

    # Configure Logging
    logger_provider = LoggerProvider(resource=resource)
    log_exporter = OTLPLogExporter(endpoint=logs_endpoint)
    logger_provider.add_log_record_processor(BatchLogRecordProcessor(log_exporter))
    set_logger_provider(logger_provider)

    # Attach OTLP handler to root logger
    handler = LoggingHandler(level=logging.INFO, logger_provider=logger_provider)
    logging.getLogger().addHandler(handler)
    logging.getLogger().setLevel(logging.INFO)

    # Manual instrumentation: explicitly configure Flask instrumentation
    # This will be applied when Flask app is created
    FlaskInstrumentor().instrument()

    # Also instrument outgoing HTTP requests if needed
    RequestsInstrumentor().instrument()

    print('=== OpenTelemetry Manual Instrumentation Configured ===')
    print(f'Service Name: {service_name}')
    print(f'Service Version: {service_version}')
    print(f'Traces Endpoint: {traces_endpoint}')
    print(f'Metrics Endpoint: {metrics_endpoint}')
    print(f'Logs Endpoint: {logs_endpoint}')
    print('=======================================================')

    return {
        'trace_provider': trace_provider,
        'meter_provider': meter_provider,
        'logger_provider': logger_provider
    }
